File: backend/app/schemas/leaderboard.py
from pydantic import BaseModel
from datetime import datetime


class LeaderboardEntryResponse(BaseModel):
    rank: int
    entity_id: str
    name: str
    creator: str | None = None
    elo: int
    winrate: float
    total_wins: int
    total_losses: int
    last_match: datetime | None = None
    badges: list[str] = []


class MyPositionResponse(BaseModel):
    rank: int
    username: str
    elo: int


class LeaderboardResponse(BaseModel):
    items: list[LeaderboardEntryResponse]
    total: int
    my_position: MyPositionResponse | None = None
    season: str
    filters: dict = {}


class SeasonListResponse(BaseModel):
    current: str
    available: list[str]

File: frontend/src/api/leaderboard.ts
import api from './client'

export interface LeaderboardItem {
  rank: number
  entity_id: string
  name: string
  elo: number
  winrate: number
  total_wins: number
  total_losses: number
  badges: string[]
  // bots only
  creator?: string
}

export interface LeaderboardResponse {
  items: LeaderboardItem[]
  total: number
  season: string
  my_position?: { rank: number; username: string; elo: number }
}

export interface SeasonsResponse {
  current: string
  available: string[]
}

export const leaderboardApi = {
  users: (season?: string) =>
    api.get<LeaderboardResponse>(`/leaderboard/users${season ? `?season=${season}` : ''}`),
  bots: (season?: string) =>
    api.get<LeaderboardResponse>(`/leaderboard/bots${season ? `?season=${season}` : ''}`),
  seasons: () => api.get<SeasonsResponse>('/leaderboard/seasons'),
}
